### PR TITLE
Feature/4088 draggable filterpane in table widget

### LINKED ISSUE #4088: [Feature] Ability to drag the filter pane in Table widget.
## Summary

As a user, I should be able to drag the filter pane on the table widget anywhere on the canvas, just like how I can do with the property pane. 

## Motivation

When the filter pane of the table widget is open it blocks and many actions and also the table data and rows, which can be a frustrating experience to the user. In order to resolve this, the filter pane can now be dragged across the canvas.

## Additional Context

[Check designs here](https://www.figma.com/file/trkiMc12VVpQBqsm7natiU/Table-Changes-Wed-7-Apr-2021?node-id=2%3A3024), 

`user flows > Table Options Dropdowns > Filter option start`

### LINKED ISSUE #5329: [Bug] Deploy Preview - Table Widget - Once when the filter is applied again user needs to click on filter icon for using other filter
## Description
Deploy Preview - https://deploy-preview-5257--appsmith-editor-ce.netlify.app/

Table Widget - It is observed that once when the filter is applied user needs to click on filter icon again to perform other filters

**Expected Result:**
Can we have a close button in the filter pane so that user can close filter manually or any appropriate behaviour

**Actual Result:**
User needs to click on filter icon each and every time for performing filter once when the filter is applied

### Steps to reproduce the behaviour:

1. Launch appmith app in the Deploy Preview https://deploy-preview-5257--appsmith-editor-ce.netlify.app/
2. Drag and Drop Table widget 
3. Click on filter option 
4. Filter pane will be displayed and filter anything and click on apply filter and observe filter pane closes

**Video Link:**
https://www.loom.com/share/6f7b2f51f9544723904bc123368f4c58

### Important Details

- Version: [14.1 (16611.1.21.161.3)]
- OS: [MacOS]
- Browser [chrome, safari]
- Environment [deploy preview]

### LINKED ISSUE #5332: [Bug]Deploy Preview - Table widget - Filter is not accessible unless if the debugger is closed when the filter is dragged and dropped to the bottom
## Description
Deploy Preview - Deploy Preview - https://deploy-preview-5257--appsmith-editor-ce.netlify.app/

It is observed that when the user drag and drops filter to the bottom of the canvas and when the debugger is opened the filter is not accessible unless the debugger is closed

**Expected Result:
Filter should be accessible even if the debugger is opened

Actual Result:**
Filter is not accessible unless if the debugger is closed

### Steps to reproduce the behaviour:

1. Launch Appmsith app in the Deploy Preview - Deploy Preview - https://deploy-preview-5257--appsmith-editor-ce.netlify.app/
2. Drag and drop table widget 
3. Open filter pane 
4. Drag the filter pane to the bottom area
5. Click on debugger
6. Click on filter icon again and observe

**Video Link:**
https://www.loom.com/share/fa56f75aca5841b482c6cce046e313b6

### Important Details

- Version: [Cloud / Self-Hosted vx.x]
- OS: [e.g.MacOSX]
- Environment [production, release, deploy preview]

### PR DESCRIPTION
**Pull request by vibhandikyash**
## Description

> As a user, I should be able to drag the filter pane on the table widget anywhere on the canvas, just like how I can do with the property pane.

Fixes #4088 
Fixes #5329
Fixes #5332

## Type of change

- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

> Tested locally.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: FEATURE/4088-draggable-filterpane-in-table-widget 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 53.92 **(0.01)** | 35.81 **(-0.04)** | 32.59 **(0.05)** | 54.5 **(0.01)**
 :green_circle: | app/client/src/actions/widgetActions.tsx | 83.02 **(1.02)** | 0 **(0)** | 50 **(3.33)** | 77.78 **(1.31)**
 :red_circle: | app/client/src/components/editorComponents/DraggableComponent.tsx | 81.97 **(-1.08)** | 58.54 **(-3)** | 73.33 **(0)** | 81.82 **(-1.2)**
 :red_circle: | app/client/src/components/editorComponents/ResizableComponent.tsx | 59.14 **(-0.2)** | 22.22 **(-0.73)** | 63.64 **(0)** | 58.24 **(-0.19)**
 :red_circle: | app/client/src/components/editorComponents/WidgetNameComponent/index.tsx | 81.13 **(-0.5)** | 63.41 **(-5.01)** | 90 **(0)** | 80.39 **(-0.46)**
 :red_circle: | app/client/src/icons/ControlIcons.tsx | 55.91 **(-0.27)** | 100 **(0)** | 8.89 **(-0.41)** | 55.91 **(-0.27)**
 :green_circle: | app/client/src/pages/Editor/GlobalHotKeys.tsx | 71.95 **(0.7)** | 39.13 **(0)** | 61.29 **(1.29)** | 70.51 **(0.77)**
 :red_circle: | app/client/src/pages/Editor/Popper.tsx | 85 **(0)** | 71.05 **(-1.17)** | 88.89 **(0)** | 85 **(0)**
 :green_circle: | app/client/src/pages/Editor/WidgetsEditor.tsx | 90.63 **(0.15)** | 71.88 **(0)** | 100 **(0)** | 90.16 **(0.16)**
 :green_circle: | app/client/src/pages/Editor/IntegrationEditor/MockDataSources.tsx | 55.56 **(10.11)** | 87.5 **(0)** | 0 **(0)** | 55.56 **(10.11)**
 :red_circle: | app/client/src/reducers/entityReducers/datasourceReducer.ts | 9.09 **(-0.17)** | 0 **(0)** | 0 **(0)** | 9.62 **(0)**
 :red_circle: | app/client/src/reducers/entityReducers/metaReducer.ts | 32.14 **(-10.72)** | 0 **(0)** | 14.29 **(-2.38)** | 32.14 **(-10.72)**
 :sparkles: :new: | **app/client/src/reducers/uiReducers/tableFilterPaneReducer.tsx** | **28.57** | **0** | **50** | **31.58**
 :red_circle: | app/client/src/sagas/WidgetOperationSagas.tsx | 44.01 **(-0.07)** | 30.28 **(0)** | 50.59 **(0)** | 44.39 **(-0.07)**
 :sparkles: :new: | **app/client/src/selectors/tableFilterSelectors.tsx** | **100** | **58.82** | **100** | **100**
 :red_circle: | app/client/src/utils/hooks/dragResizeHooks.tsx | 75.93 **(-4.07)** | 22.22 **(-17.78)** | 60 **(-1.54)** | 75 **(-3.38)**
 :green_circle: | app/client/src/utils/hooks/useClickOpenPropPane.tsx | 81.82 **(0.69)** | 53.85 **(1.22)** | 100 **(0)** | 80.77 **(0.77)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@vibhandikyash too many conflicts here, Kindly resolve

**Review by somangshu — CHANGES_REQUESTED** [review]
- The filter-pane should load inside the table view on the first load
- On click of apply do not close the filter-pane

**Comment by somangshu** [comment]
@vivekverma2312 please leave your comments here. 

@vibhandikyash lets wait for the feedback and then pick everything up together

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash @somangshu Tested Table filter pane drag and drop functionality and raised the below bugs

#5329 and #5332

**Comment by vibhandikyash** [comment]
> @vibhandikyash @somangshu Tested Table filter pane drag and drop functionality and raised the below bugs
> 
> #5329 and #5332
@YogeshJayaseelan  I've fixed both of these issues(#5329 and #5332) and committed them here.

File: app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Binding/Widgets_form_input_table_default_validation.js [hunk 1]
```diff
@@ -33,10 +33,10 @@ describe("Binding the multiple input Widget", function() {
         .first()
         .invoke("attr", "value")
         .should("contain", tabValue);
-//       cy.get(publish.inputWidget + " " + "input")
-//         .last()
-//         .invoke("attr", "value")
-//         .should("contain", tabValue);
+      //       cy.get(publish.inputWidget + " " + "input")
```
Comment (somangshu) [inline app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Binding/Widgets_form_input_table_default_validation.js:36]: Please remove this if not required!

Comment (somangshu) [reply app/client/cypress/integration/Smoke_TestSuite/ClientSideTests/Binding/Widgets_form_input_table_default_validation.js:36]: Need to resolve the conflict as well

**Comment by somangshu** [comment]
@vibhandikyash please link additional issues fixed under the same PR

**Comment by somangshu** [comment]
@vibhandikyash resolve the conflicts

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash Verified bug fixes 5332 and 5329 and they are working as expected.

**Review by somangshu — DISMISSED** [review]

**Review by somangshu — APPROVED** [review]

**Comment by marks0351** [comment]
Screenshot 2021-07-07 at 12 31 41 PM
The move icon covers the help text. can u pls check @vibhandikyash

**Comment by marks0351** [comment]
I have reviewed the code @vibhandikyash it looks great. 

Code solution feedback:
I think that having the tab filter pane movements on the app state might not be recommended. Do you think we can handle all this within the table widget by creating an internal state instead?

Bugs I found: 
 - Help text overlap which I have commented
 - Not able to select any other widget when table filter pane is open.

also cud u get the filter utils part reviewed from @riodeuno 
pls do resolve the conflicts as well.

**Review by marks0351 — CHANGES_REQUESTED** [review]
Added review comments, pls check.

**Comment by marks0351** [comment]
@vibhandikyash pls do add jest and cypress tests validating the feature.

**Comment by vivekverma2312** [comment]
> Screenshot 2021-07-07 at 12 31 41 PM
> 
> The move icon covers the help text. can u pls check @vibhandikyash

Looked at the loom video of issues #5329 & #5332 reported by yogesh and observed that this was working fine when @YogeshJayaseelan was testing this. Can we figure out how this issue got introduced and is there any way we can catch such issues which got introduced after giving signoff.
cc: @somangshu

**Comment by somangshu** [comment]
@vivekverma2312 the change probably was introduced because of the a merge conflict resolution. @vibhandikyash we need to verify this

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@vicky-primathon need your help to fix failing test and merge conflict, we need to close this, @vibhandikyash is not available for a couple weeks.

**Comment by vicky-primathon** [comment]
Continuing the feature here https://github.com/appsmithorg/appsmith/pull/5886

**Comment by YogeshJayaseelan** [comment]
@vibhandikyash @vicky-primathon @somangshu 
Verified the fixes and the following issues are working as expected

1) #5329 - Verified
2) #5332 - Verified
3) Help text overlap - Verified
4) Not able to select(by clicking on the widget) any other widget when table filter pane is open - Verified

**New issue:**
5) Filter functionality is not working as expected - ID ends with 1 or First-name ends with e

For the above table data use the following sample api and bind with the table data - https://reqres.in/api/users?page=2

https://loom.com/share/1514c5e725b343c290e65b82cd478f6c

6) Can we have clear button inside the filter pane so that without removing the existing filter user can clear the data

**Comment by somangshu** [comment]
@YogeshJayaseelan please continue the conversation in the main/new thread/PR

**Pull request closed without merging**